src/uct/ib/rdmacm/rdmacm_cm.c

```diff
@@ -242,60 +238,59 @@ static void uct_rdmacm_cm_handle_event_connect_response(struct rdma_cm_event *ev
     remote_data.dev_addr_length   = addr_length;
 
     uct_rdmacm_cm_ep_client_connect_cb(cep, &remote_data, hdr->status);
+    cep->flags |= UCT_RDMACM_CM_EP_CONNECTED;
 
     ucs_free(dev_addr);
 
     if (rdma_establish(event->id)) {
         ucs_error("rdma_establish on ep %p (to server addr=%s) failed: %m",
                   cep, ucs_sockaddr_str(remote_addr, ip_port_str,
                                         UCS_SOCKADDR_STRING_LEN));
-        /* in case of an error here, call disconnect because the client already
-         * called the connect_cb with status UCS_OK so the client is considered
-         * as connected to the server at this point. */
-        cep->disconnect_cb(&cep->super.super, cep->user_data);
-    }
-}
-
-static void uct_rdmacm_cm_handle_event_rejected(struct rdma_cm_event *event)
-{
-    uct_rdmacm_cm_ep_t         *cep = event->id->context;
-    uct_rdmacm_priv_data_hdr_t *hdr;
-    uct_cm_remote_data_t       remote_data;
-    ucs_status_t               status;
-
-    ucs_assert(event->id == cep->id);
-
-    /* Network reject or called by rdma_reject on the server */
-    remote_data.field_mask = 0;
-    if (event->param.conn.private_data != NULL) {
-        /* The server always passes private data when rejecting a connection request */
-        hdr    = (uct_rdmacm_priv_data_hdr_t *)event->param.conn.private_data;
-        status = hdr->status;
-        ucs_assert(hdr->length == 0);
-    } else {
-        /* Network reject */
-        status = UCS_ERR_REJECTED;
+        uct_rdmacm_cm_ep_error_cb(cep, &remote_data, status);
     }
-
-    ucs_debug("rdmacm event rejected on ep %p (cm_id %p) with status %s.",
-              cep, event->id, ucs_status_string(status));
-
-    uct_rdmacm_cm_ep_client_connect_cb(cep, &remote_data, status);
 }
 
 static void uct_rdmacm_cm_handle_event_established(struct rdma_cm_event *event)
 {
     uct_rdmacm_cm_ep_t *cep = event->id->context;
 
     ucs_assert(event->id == cep->id);
-    cep->wireup.server.connect_cb(&cep->super.super, cep->user_data, UCS_OK);
+    uct_rdmacm_cm_ep_server_connect_cb(cep, UCS_OK);
+    cep->flags |= UCT_RDMACM_CM_EP_CONNECTED;
 }
 
 static void uct_rdmacm_cm_handle_event_disconnected(struct rdma_cm_event *event)
 {
-    uct_rdmacm_cm_ep_t *cep = event->id->context;
+    uct_rdmacm_cm_ep_t *cep      = event->id->context;
+    struct sockaddr *remote_addr = rdma_get_peer_addr(event->id);
+    char            ip_port_str[UCS_SOCKADDR_STRING_LEN];
+
+    ucs_debug("%s received disconnect event. status = %d. Peer: %s.",
```

Comment:
"%s got disconnect event, status %d peer %s"
BTW we already have debug print in https://github.com/openucx/ucx/pull/4084/files#diff-f3bf293c768549f806213fabfa19865bR303, why need another one?

Reply:
@yosefe because only here i have a valid ep so i can tell if it's client or server. 
(ep for client is always valid in uct_rdmacm_cm_process_event but not for server)

Reply:
ok